File: packages/javascript/tests/core/headers.test.ts
import { describe, it, expect } from 'vitest'
import { buildHeaders, daysUntilSunset } from '../../src/core/headers'
import type { EndpointConfig } from '../../src/core/models'

function ep(overrides: Partial<EndpointConfig> = {}): EndpointConfig {
  return {
    path: '/api/v1/users',
    methods: ['GET'],
    deprecated_at: '2025-01-01',
    sunset_at: null,
    ...overrides,
  }
}

describe('buildHeaders', () => {
  it('always includes Deprecation header', () => {
    const h = buildHeaders(ep())
    expect(h.Deprecation).toContain('2025-01-01')
  })

  it('includes Sunset header when sunset_at is set', () => {
    expect(buildHeaders(ep({ sunset_at: '2026-01-01' })).Sunset).toBe('2026-01-01')
  })

  it('omits Sunset header when sunset_at is null', () => {
    expect(buildHeaders(ep()).Sunset).toBeUndefined()
  })

  it('includes Link header with successor-version rel', () => {
    const h = buildHeaders(ep({ successor: '/api/v2/users' }))
    expect(h.Link).toContain('successor-version')
    expect(h.Link).toContain('/api/v2/users')
  })

  it('includes Link header with deprecation rel for migration_doc', () => {
    const h = buildHeaders(ep({ migration_doc: 'https://docs.example.com' }))
    expect(h.Link).toContain('rel="deprecation"')
  })

  it('combines multiple Link entries', () => {
    const h = buildHeaders(ep({ successor: '/api/v2/users', migration_doc: 'https://docs.example.com' }))
    expect(h.Link?.split(', ').length).toBe(2)
  })
})

describe('daysUntilSunset', () => {
  it('returns null when no sunset_at', () => {
    expect(daysUntilSunset(ep())).toBeNull()
  })

  it('returns positive number for future sunset', () => {
    expect(daysUntilSunset(ep({ sunset_at: '2099-12-31' }))).toBeGreaterThan(0)
  })

  it('returns negative number for past sunset', () => {
    expect(daysUntilSunset(ep({ sunset_at: '2000-01-01' }))).toBeLessThan(0)
  })
})

File: packages/go/core/headers_test.go
package core

import (
	"strings"
	"testing"
)

func strPtr(s string) *string { return &s }

func TestBuildHeaders_Deprecation(t *testing.T) {
	ep := EndpointConfig{Path: "/api/v1/users", Methods: []string{"GET"}, DeprecatedAt: "2025-01-01"}
	h := BuildHeaders(ep)
	if h["Deprecation"] != "@1735689600" {
		t.Fatalf("expected Deprecation=@1735689600, got %q", h["Deprecation"])
	}
}

func TestBuildHeaders_Sunset(t *testing.T) {
	ep := EndpointConfig{
		Path:         "/api/v1/users",
		Methods:      []string{"GET"},
		DeprecatedAt: "2025-01-01",
		SunsetAt:     strPtr("2026-06-01"),
	}
	h := BuildHeaders(ep)
	if h["Sunset"] != "Mon, 01 Jun 2026 00:00:00 GMT" {
		t.Fatalf("expected Sunset=Mon, 01 Jun 2026 00:00:00 GMT, got %q", h["Sunset"])
	}
}

func TestBuildHeaders_LinkSuccessor(t *testing.T) {
	ep := EndpointConfig{
		Path:         "/api/v1/users",
		Methods:      []string{"GET"},
		DeprecatedAt: "2025-01-01",
		Successor:    strPtr("/api/v2/users"),
	}
	h := BuildHeaders(ep)
	if !strings.Contains(h["Link"], "successor-version") {
		t.Fatalf("expected Link with successor-version, got %q", h["Link"])
	}
}

func TestBuildHeaders_LinkMigrationDoc(t *testing.T) {
	ep := EndpointConfig{
		Path:         "/api/v1/users",
		Methods:      []string{"GET"},
		DeprecatedAt: "2025-01-01",
		MigrationDoc: strPtr("https://docs.example.com/migration"),
	}
	h := BuildHeaders(ep)
	if !strings.Contains(h["Link"], "deprecation") {
		t.Fatalf("expected Link with deprecation rel, got %q", h["Link"])
	}
}

func TestDaysUntilSunset_NoSunset(t *testing.T) {
	ep := EndpointConfig{Path: "/api/v1/users", DeprecatedAt: "2025-01-01"}
	if DaysUntilSunset(ep) != nil {
		t.Fatal("expected nil when no sunset")
	}
}

func TestDaysUntilSunset_FutureSunset(t *testing.T) {
	ep := EndpointConfig{
		Path:         "/api/v1/users",
		DeprecatedAt: "2025-01-01",
		SunsetAt:     strPtr("2099-01-01"),
	}
	days := DaysUntilSunset(ep)
	if days == nil || *days <= 0 {
		t.Fatal("expected positive days for future sunset")
	}
}

func TestDaysUntilSunset_PastSunset(t *testing.T) {
	ep := EndpointConfig{
		Path:         "/api/v1/users",
		DeprecatedAt: "2025-01-01",
		SunsetAt:     strPtr("2020-01-01"),
	}
	days := DaysUntilSunset(ep)
	if days == nil || *days >= 0 {
		t.Fatal("expected negative days for past sunset")
	}
}
